### DISCUSSION TITLE
Terragrunt private Terraform registries hostname are transform to "registry.terraform.io"

### DISCUSSION THREAD

**Discussion by ismtabo** [renovatebot/renovate#26644] [Request Help] [answered]
### How are you running Renovate?

Using GitHub Mend Renovate service.

### If you're self-hosting Renovate, select which platform you are using.

None

### Was this something which used to work for you, and then stopped?

None

### Describe the problem

In a private repository, we used to have Terragrunt dependencies using source links to a GitHub repository using a URL like "`github.com/:org/:repo//:path-to-module?...`". Recently, we changed to using a private Terraform Registry, and now these source links seem like "`tfr://registry.hostname.com/:org/:name/:cloud`".

Renovate shows a warning about problems fetching the dependencies. We have checked the logs, and they show requests to `registry.terraform.io` instead of the hostname of the registry present in the Terragurnt files.

We have searched what may have caused the problem and detected that the terragrunt data provider extracts the dependency based on the `tfr://host.name.es/...` pattern, and the registry group inside the regex is not used later in the analysis of dependencies. See links below:

Terragrunt's Terraform Registry source link regex:
https://github.com/renovatebot/renovate/blob/main/lib/modules/manager/terragrunt/modules.ts#L16-L18

Terragrunt's Terraform Registry source link analysis:
https://github.com/renovatebot/renovate/blob/main/lib/modules/manager/terragrunt/modules.ts#L65-L74

On the other hand, these dependencies work correctly with the `terraform` data provider, resolving the registry hostname to the proper one.

### How to reproduce the error

In a GitHub repository with Renovate configured and running. Add a `terragrunt.hcl` file with the following content:

```hcl
terraform {
  source = "tfr://registry.domain.com/miguelhrocha/helloworld/aws?version=1.0.0"
}
```

The Terragrunt file contains a dependency to the module `miguelhrocha/helloworld/aws` with version `1.0.0` published at private registry at `registry.domain.com`.

#### Expected behavior

Renovate request the private Terraform registry to list Terraform module versions, and **fails** due to the registry host **does not exist**.

#### Actual behavior

Renovate **does request** `registry.terraform.io` host instead, **resolving** the module to https://registry.terraform.io/modules/miguelhrocha/helloworld/aws/latest.

### Relevant debug logs

<details><summary>Logs</summary>

Example of failed request to `registry.terraform.io` instead of registry host.
```
DEBUG: GET https://registry.terraform.io/v1/modules/***/***/*** = (code=ERR_NON_2XX_3XX_RESPONSE, statusCode=404 retryCount=0, duration=57)
DEBUG: Datasource 404
DEBUG: Failed to look up terraform-module package ***/***/***
```

</details>

### Have you created a minimal reproduction repository?

Yes, the minimal reproduction repository is available in the following link:
https://github.com/ismtabo/renovate-bug-resolving-private-terraform-registries-terragrunt-files

**Comment by rarkins** [comment]
Why do you need a registry to reproduce if Renovate is not looking up that registry anyway? You need to reproduce the _problem_, not reproduce the eventual solution

**Reply by ismtabo** [reply]
The description tries to settle the requirements need to reproduce the error, let me rephrase it to try to make it more clear.

**Reply by rarkins** [reply]
If you want to reproduce the problem of renovate looking up the wrong server then I don't think you need to be running the right server. You just need the terraform file in the repo

**Reply by ismtabo** [reply]
Yeah, you are right, I didn't understand your point. It is enough just with a terragrunt file. I'll update it.

**Reply by rarkins** [reply]
Please put it in a reproduction repository and run the bot against it to confirm it reproduces the error

**Reply by ismtabo** [reply]
Sure, I added a reproduction repository to the discussion description. Thanks!

**Reply by rarkins** [reply]
I think that it's enough to simply run with `--dry-run=extract` and observe the results:

```
 INFO: Extracted dependencies (repository=ismtabo/renovate-bug-resolving-private-terraform-registries-terragrunt-files)
       "packageFiles": {
         "terragrunt": [
           {
             "deps": [
               {
                 "depType": "terragrunt",
                 "depName": "miguelhrocha/helloworld/aws",
                 "currentValue": "1.0.0",
                 "datasource": "terraform-module"
               }
             ],
             "packageFile": "terragrunt.hcl"
           }
         ]
       }
```

The registry url is missing here

**Reply by marsou77** [reply]
Hello, I'have the same issue on the latest renovate version. But it is working fine with `ghcr.io/renovatebot/renovate:37.74.1`

**Reply by rarkins** [reply]
I don't think this ever worked

**Reply by marsou77** [reply]
I just got it working with the tag I given. But with this tag I don't even configure `hostRules` or something. It just works. Waiting the fix :) Thank you by the way!

**Comment by rarkins** [comment, marked as answer]
Converted to issue: https://github.com/renovatebot/renovate/issues/26668
